Git Client Support On-Prem Github

midnightconman opened the pull request:
This change adds a `git-endpoint` flag for passing to git client. This is an alternative implementation of 
PR https://github.com/kubernetes/test-infra/pull/10129

components: hook, tide and the cherrypicker external plugin.

Fixes #10068, #11014

/assign @fejta @stevekuznetsov

I welcome testing suggestions :smile:

stevekuznetsov commented:
/ok-to-test

stevekuznetsov commented on prow/git/git.go:
```diff
@@ -50,7 +49,7 @@ type Client struct {
-	// base is the base path for git clone calls. For users it will be set to
+	// bases are the base paths for git clone calls. For users it will be set to
 	// GitHub, but for tests set it to a directory with git repos.
 	base string
```
did you mean to make this a slice?

midnightconman replied:
Yep, on another branch I have been testing multiple gitEndpoints. Nothing promising yet though.

midnightconman replied:
I'll revert this comment for now.

stevekuznetsov commented on prow/git/git.go:
```diff
@@ -138,7 +137,17 @@ func (c *Client) Clone(repo string) (*Repo, error) {
+			return nil, err
+		}
+
+		nu := &url.URL{
```
Could we use more descriptive variable names than `nu` please?

fejta replied:
Can't we just reuse `u`?

midnightconman replied:
Yes and Yes... please see new commit.

stevekuznetsov commented on prow/flagutil/github.go:
```diff
@@ -31,6 +31,7 @@ import (
 // GitHubOptions holds options for interacting with GitHub.
 type GitHubOptions struct {
 	endpoint            Strings
+	gitEndpoint         string
```
Is it valid for `gitEndpoint` and `endpoint` to be on two different domains?

midnightconman replied:
Yes, here people can put whatever they want.

Do you have an idea on validation or enforcement of this?

yastij commented on prow/flagutil/github.go:
```diff
@@ -102,7 +105,7 @@ func (o *GitHubOptions) GitHubClient(secretAgent *secret.Agent, dryRun bool) (cl
 // GitClient returns a Git client.
 func (o *GitHubOptions) GitClient(secretAgent *secret.Agent, dryRun bool) (client *git.Client, err error) {
-	client, err = git.NewClient()
+	client, err = git.NewClient(o.gitEndpoint)
```
This won't work, especially if you are passing endpoints like `api.github.com` and `github.mycorp/api/v3`

yastij replied:
Sorry, mis-read, I thought you were re-using the github endpoint

midnightconman commented:
> Would it make sense to make prow's git client agnostic to github?
> 
> /lgtm

I thought about calling this flag `scm-endpoint` and creating a new flag called `scm-api-endpoint` which would be the same value as `github-endpoint` until we deprecate it... thoughts?

fejta commented:
I like this approach. Any chance you can coordinate with @yastij to merge #10129 with a similar strategy?

midnightconman commented:
Added a test for an example git-endpoint and fixed git test package naming.

@yastij Does this approach work for you? Is there anything we can add to this to match your use case?

yastij commented on prow/flagutil/github.go:
```diff
@@ -49,7 +50,9 @@ func (o *GitHubOptions) AddFlagsWithoutDefaultGithubTokenPath(fs *flag.FlagSet)
 	o.endpoint = NewStrings("https://api.github.com")
+	o.gitEndpoint = "https://github.com"
 	fs.Var(&o.endpoint, "github-endpoint", "GitHub's API endpoint (may differ for enterprise).")
+	fs.StringVar(&o.gitEndpoint, "git-endpoint", o.gitEndpoint, "GitHub endpoint (may differ for enterprise).")
```
@midnightconman - can you add a test for the validation of github options ?

midnightconman replied:
Great idea! Done.

yastij commented:
/lgtm
/hold

feel free to cancel the hold, once you add testing for the validation.

fejta commented on prow/flagutil/github.go:
```diff
@@ -49,7 +50,9 @@ func (o *GitHubOptions) AddFlagsWithoutDefaultGithubTokenPath(fs *flag.FlagSet)
 
 func (o *GitHubOptions) addFlags(wantDefaultGithubTokenPath bool, fs *flag.FlagSet) {
 	o.endpoint = NewStrings("https://api.github.com")
+	o.gitEndpoint = "https://github.com"
```
Just put the string literal into the `StringVar()` call (the third arg is the default value), like we do for github-token-path.

We have an `o.endpoint = NewStrings()` because `Var()` doesn't include a default value param.

midnightconman replied:
Makes sense... done.

yastij commented on prow/flagutil/github.go:
```diff
@@ -66,6 +68,10 @@ func (o *GitHubOptions) Validate(dryRun bool) error {
 		}
 	}
 
+	if _, err := url.ParseRequestURI(o.gitEndpoint); err != nil {
```
Can you also add test file to test validation of github options ?

midnightconman replied:
Sure :smile:

yastij commented:
/retest

Can you squash your commits ? Thanks !

midnightconman commented:
@fejta @stevekuznetsov - Can I get one last look at this?

/hold cancel

stevekuznetsov commented on prow/flagutil/github_test.go:
```diff
@@ -0,0 +1,78 @@
+		expectedErr bool
+	}{
+		{
+			name:        "no token provided",
```
It is often nice to have the cause and effect here together:

```
"no token provided is ok"
```

midnightconman replied:
No problem, updated... let me know if we need more information.

stevekuznetsov commented:
/hold cancel

yastij commented:
/lgtm
/assign @fejta 
Thanks

midnightconman commented on prow/git/git.go:
```diff
@@ -288,9 +296,17 @@ func (r *Repo) Push(repo, branch string) error {
+		return err
+	}
+	gitURL.User = url.UserPassword(r.user, r.pass)
+	gitURL.Path = strings.Join([]string{r.user, repo}, "/")
```
Sorry guys, this should be the last change... I found that this was removed during reuse of the `u` variable, so wanted to get this in so we didn't have issues.

Please review and let me know if we need anything else before merge

stevekuznetsov replied:
Did this fail any tests when it was missing? If not, can we add a test that would have failed?

midnightconman replied:
Can we make this a future enhancement? There isn't any existing testing for `git.Push` and the function seems it would require re-work to support mock testing.

stevekuznetsov replied:
Sure -- I would suggest making the implementation of `func (r *Repo) gitCommand(arg ...string)` use a member of `type Repo struct` and injecting that during test time to record the actions that would be taken, that would allow us to test the underlying git call for input args to the client from Go.

fejta replied:
Ideally, unit tests to validate new functionality should come in the same PR as the new functionality. We should consider this a future enhancement.

My concern is that we LGTM'd (and therefore almost merged) something which would have caused an outage and no test failed to catch that scenario.

This suggests we are missing useful test coverage, an easy way to test this would be to encapsulate `gitURL(r Repo) (string, error)` and then do
```golang
cases := []struct{
  name string
  input Repo
  expected string
  err bool
}{
   { name: "no user returns error",
     err: true,
   },
  { // etc
  }
}

for _, tc := range cases {
  actual, err := gitRepo(tc.Repo)
  if actual != tc.expected { tc.Errorf() }
}
```

Maybe gitRemote is a better name?

Then this function becomes pretty trivial:
```golang
remote, err := gitRemote(repo)
return r.gitCommand("push", remote, branch).CombinedOutput()
```

which I don't think requires further unit testing

fejta commented on prow/git/git.go:
```diff
@@ -288,9 +296,17 @@ func (r *Repo) Push(repo, branch string) error {
+
+	gitURL, err := url.Parse(r.base)
+	if err != nil {
+		return err
```
`Push()` can fail in three ways:
* User not provided
* Base fails to parse
* Push command failed

I would prefer we do something like `fmt.Errorf("parse base: %v", err)` to help distinguish things here rather than just pass the err through in each scenario

midnightconman replied:
Added a more descriptive error message.

midnightconman commented:
/hold

Let me take another stab at this.

On Mon, Feb 4, 2019, 19:35 Kubernetes Prow Robot <notifications@github.com
wrote:

> New changes are detected. LGTM label has been removed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/11075#issuecomment-460501494>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABp-O-hOMvy6zpnwl9GElT-0rfhgOj2Cks5vKPvogaJpZM4adccF>
> .
>

fejta commented on prow/git/git.go:
```diff
@@ -316,6 +334,17 @@ func (r *Repo) Config(key, value string) error {
 }
 
+// GitRemote builds a remote url from user, pasword, and a slice of path items
+func GitRemote(base string, user string, pass string, repo string) (*url.URL, error) {
```
Why do you want to export this method? If you don't want to (don't think its needed), please don't capitalize the method.

midnightconman replied:
I am exporting this so we can test it, `git_test.go` is a different package then `git.go`, so we need to export functions to test them... not sure why they are that way though 🤷‍♂️

fejta replied:
These two files are in the same package (git)

fejta replied:
Huh, well they both ought to be in the same package. Not sure why the test is in git_test rather than git.

fejta replied:
Huh, well they both ought to be in the same package. Not sure why the test is in git_test rather than git.

midnightconman replied:
I think this is because of the way local_git is implemented. Can we leave this git/git_test re-work off this PR? It seems like it should have it's own re-work PR.

midnightconman commented:
/hold cancel

On Mon, Feb 4, 2019, 23:40 Kubernetes Prow Robot <notifications@github.com
wrote:

> Thanks for your pull request. Before we can look at your pull request,
> you'll need to sign a Contributor License Agreement (CLA).
>
> 📝 *Please follow instructions at
> https://git.k8s.io/community/CLA.md#the-contributor-license-agreement
> <https://git.k8s.io/community/CLA.md#the-contributor-license-agreement> to
> sign the CLA.*
>
> It may take a couple minutes for the CLA signature to be fully registered;
> after that, please reply here with a new comment and we'll verify. Thanks.
> ------------------------------
>
>    - If you've already signed a CLA, it's possible we don't have your
>    GitHub username or you're using a different email address. Check your
>    existing CLA data and verify that your email is set on your git commits
>    <https://help.github.com/articles/setting-your-email-in-git/>.
>    - If you signed the CLA as a corporation, please sign in with your
>    organization's credentials at
>    https://identity.linuxfoundation.org/projects/cncf to be authorized.
>    - If you have done the above and are still having issues with the CLA
>    being reported as unsigned, please email the CNCF helpdesk:
>    helpdesk@rt.linuxfoundation.org
>
> Instructions for interacting with me using PR comments are available here
> <https://git.k8s.io/community/contributors/guide/pull-requests.md>. If
> you have questions or suggestions related to my behavior, please file an
> issue against the kubernetes/test-infra
> <https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:>
> repository. I understand the commands that are listed here
> <https://go.k8s.io/bot-commands>.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/11075#issuecomment-460541177>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABp-O46PfwVmJNQa7aJ4BVF6Hhv5Rk5Pks5vKTVbgaJpZM4adccF>
> .
>

midnightconman commented on prow/git/git.go:
```diff
@@ -109,6 +108,10 @@ func (c *Client) getCredentials() (string, string) {
 	return c.user, string(c.tokenGenerator())
 }
 
+func (c *Client) GetBase() string {
```
Added comments to `GetBase` function.

midnightconman commented on prow/git/git.go:
```diff
@@ -316,6 +334,17 @@ func (r *Repo) Config(key, value string) error {
 }
 
+// GitRemote builds a remote url from user, pasword, and a slice of path items
+func GitRemote(base string, user string, pass string, repo string) (*url.URL, error) {
```
Renamed function to `Remote`.

midnightconman commented:
Okie dokie, added `Remote` function and tests. Also added an in-function path validation to `Push`, so an error will be generated if signatures change... please take another look and let me know if I can do anything else. 😄 Once you guys are happy I'll rebase these into one commit.

@fejta @stevekuznetsov

fejta commented on prow/git/git.go:
```diff
@@ -316,6 +334,17 @@ func (r *Repo) Config(key, value string) error {
+func Remote(base string, user string, pass string, pathItems ...string) (*url.URL, error) {
+	newURL, err := url.Parse(base)
+	if err != nil {
+		return &url.URL{}, fmt.Errorf("Error while parsing base: %v", err)
```
Use nil rather than a default URL

midnightconman replied:
Sure, no problem... Done.

fejta commented:
/lgtm
/hold cancel

yastij commented:
/lgtm

midnightconman commented:
@*stevekuznetsov Can you take one last look at this, please? *

On Tue, Feb 5, 2019, 14:53 Yassine TIJANI <notifications@github.com wrote:

> /lgtm
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kubernetes/test-infra/pull/11075#issuecomment-460836676>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABp-O9wRLA2y1mWxemRXlbqzSIZovoSTks5vKgtsgaJpZM4adccF>
> .
>

stevekuznetsov commented on prow/git/git.go:
```diff
@@ -288,9 +296,19 @@ func (r *Repo) Push(repo, branch string) error {
+
+	// We can't push without user and repo in remote path
+	if !strings.Contains(remote.String(), strings.Join([]string{r.user, repo}, "/")) {
+		return errors.New("git remote must contain :user and :repo to push")
```
Shouldn't this be validated at flag parse time? Why bother creating a `git` client that can't push?

midnightconman replied:
I can remove this, but I put it in here to prevent weird silent errors, that could be introduced by flag logic changes.

It's up to you guys, leave it in, take it out... I am good either way.

fejta replied:
Steve what is your concern with doing this validation?

I can see the case for being too lazy to add it, but why ask to remove it?

stevekuznetsov replied:
It just seems crazy to me. Like why would you validate this at flag load time, then call a func with it, then sanity-check that the func did what you wanted? It's just muddying the code. We should write code that validates up front and uses types where possible to make it impossible to error later.

midnightconman replied:
What's the verdict here, leave it in, or take it out?

stevekuznetsov replied:
I believe it would be better for our code health to use the type system here to make sure that if we validate this upfront, it is not possible for the code downstream of it to error, and that would let us not need these runtime sanity checks

midnightconman replied:
Deal... let me do a commit.

stevekuznetsov commented on prow/git/git.go:
```diff
@@ -316,6 +334,17 @@ func (r *Repo) Config(key, value string) error {
 
+// Remote builds a remote url from user, pasword, and a slice of path items
+func Remote(base string, user string, pass string, pathItems ...string) (*url.URL, error) {
+	newURL, err := url.Parse(base)
```
Don't we parse this at flag load time? Why don't we store the URL object and not need to do this again later?

midnightconman replied:
That sounds like a good idea, but do we have to re-implement all these functions in this PR? This isn't going to be the last time these files are touch for a while and it might even make sense in the graphql change PR I have behind this one.

Can we get this merged and work on re-factoring and re-implementing over time? This PR is already size/L and all we did was add a flag :wink:

midnightconman replied:
I also went ahead and created an enhancement issue for this: https://github.com/kubernetes/test-infra/issues/11166

midnightconman replied:
Are we blocking approval on re-working signatures to use `*url.URL`, or can we save that for another PR?

stevekuznetsov replied:
I would like to, yes.

midnightconman commented:
/hold

I have run `bazel test` for all of prow, no failed tests... I plan to functionally test this tomorrow to make sure there are no gitClient issues. When I have verified this is working correctly, I'll remove hold. 

@fejta @stevekuznetsov Please have a look at this and let me know if I can do anything else. Thank you so much for the PR reviews and comments, I just started using prow in my company and we are really happy with what we have so far.

stevekuznetsov commented on prow/git/git.go:
```diff
@@ -288,16 +296,22 @@ func (r *Repo) Push(repo, branch string) error {
 func (r *Repo) CheckoutPullRequest(number int) error {
 	r.logger.Infof("Fetching and checking out %s#%d.", r.repo, number)
-	if b, err := retryCmd(r.logger, r.Dir, r.git, "fetch", r.base+"/"+r.repo, fmt.Sprintf("pull/%d/head:pull%d", number, number)); err != nil {
+	//r.base.Path = strings.Join([]string{r.base.Path, r.repo, fmt.Sprintf("pull/%d/head:pull%d", number, number)}, "/")
```
What's going on here?

midnightconman replied:
Good catch, removed rework comment.

Thanks Steve :smile:

yastij commented:
/lgtm
/hold cancel

midnightconman commented:
I functionally verified this on my end, couldn't find any issues.

apesternikov commented on prow/git/git.go:
```diff
@@ -147,7 +155,7 @@ func (c *Client) Clone(repo string) (*Repo, error) {
 			return nil, err
 		}
-		remote := fmt.Sprintf("%s/%s", base, repo)
+		remote := fmt.Sprintf("%s/%s", base.Path, repo)
```
initial clone is failing because of this, fix is here: https://github.com/kubernetes/test-infra/pull/11214

apesternikov commented:
fixes https://github.com/kubernetes/test-infra/issues/11219

stevekuznetsov commented on prow/git/git.go:
```diff
@@ -138,7 +142,11 @@ func (c *Client) Clone(repo string) (*Repo, error) {
 	base := c.base
 	user, pass := c.getCredentials()
 	if user != "" && pass != "" {
```
Yuck. This is extant code that _only_ handles testing cases and totally makes this PR not tested by any of the normal means

stevekuznetsov replied:
We should not allow bad values like that used for testing in production clients

spiffxp replied:
if only we had some way of knowing which lines of code were covered by our tests

cjwagner commented:
This PR has been reverted by https://github.com/kubernetes/test-infra/pull/11262 

@midnightconman @stevekuznetsov Can we wait for https://github.com/kubernetes/test-infra/pull/11248 before reimplementing this?
